// @webaudiomodules/sdk-parammgr/src/ParamMgrFactory.js
import addFunctionModule from 'webaudiomodules/sdk-parammgr/src/sdk/src/addFunctionModule.js';
import processor from 'webaudiomodules/sdk-parammgr/src/ParamMgrProcessor.js';
import ParamMappingConfigurator from 'webaudiomodules/sdk-parammgr/src/ParamConfigurator.js';
import ParamMgrNode from 'webaudiomodules/sdk-parammgr/src/ParamMgrNode.js';
/** @typedef {import('@webaudiomodules/api').WebAudioModule} WebAudioModule */
/** @typedef {import('webaudiomodules/sdk-parammgr/src/types.js').ParametersMappingConfiguratorOptions} ParametersMappingConfiguratorOptions */
/** @typedef {import('webaudiomodules/sdk-parammgr/src/types.js').ParamMgrOptions} ParamMgrOptions */
/** @typedef {import('webaudiomodules/sdk-parammgr/src/types.js').AudioWorkletRegister} AudioWorkletRegister */

export default class ParamMgrFactory {
	/**
	 * @param {WebAudioModule} module
	 * @param {ParametersMappingConfiguratorOptions} [optionsIn = {}]
	 */
	static async create(module, optionsIn = {}) {
		const { audioContext, moduleId } = module;
		const instanceId = optionsIn.instanceId || module.instanceId;
		const groupId = optionsIn.groupId || module.groupId;
		const { paramsConfig, paramsMapping, internalParamsConfig } = new ParamMappingConfigurator(optionsIn);
		const initialParamsValue = Object.entries(paramsConfig)
			.reduce((currentParams, [name, { defaultValue }]) => {
				currentParams[name] = defaultValue;
				return currentParams;
			}, {});
		const serializableParamsConfig = Object.entries(paramsConfig)
			.reduce((currentParams, [name, { id, label, type, defaultValue, minValue, maxValue, discreteStep, exponent, choices, units }]) => {
				currentParams[name] = { id, label, type, defaultValue, minValue, maxValue, discreteStep, exponent, choices, units };
				return currentParams;
			}, {});
		await addFunctionModule(audioContext.audioWorklet, processor, moduleId, serializableParamsConfig);
		/** @type {ParamMgrOptions} */
		const options = {
			internalParamsConfig,
			parameterData: initialParamsValue,
			processorOptions: {
				paramsConfig,
				paramsMapping,
				internalParamsMinValues: Object.values(internalParamsConfig)
					.map((config) => Math.max(0, config?.minValue || 0)),
				internalParams: Object.keys(internalParamsConfig),
				groupId,
				instanceId,
				moduleId,
			},
		};
		const node = new ParamMgrNode(module, options);
		await node.initialize();
		return node;
	}
}
